### PR TITLE
[glsl-in] Implement `MathFunction::Clamp`

### PR DESCRIPTION
**Pull request by sjinno**
As the title says, added the implementation of `MathFunction::Clamp`. Feedback is very much appreciated! :)

### CODE REVIEW

File: src/front/glsl/constants.rs
Comment (jimblandy) [inline src/front/glsl/constants.rs:266]: Generally, Naga IR follows the WGSL spec, and that says that the `clamp` function operates "component-wise when [the argument type] is a vector". It's fine not to bother handling vector arguments for now, but this should be a `NotImplemented` error, not an `InvalidMathArg` error.

File: src/front/glsl/constants.rs
Comment (jimblandy) [inline src/front/glsl/constants.rs:276]: Rather than repeat this code for every `MathFunction` variant, it would be better to pull it out of the `match fun` so that all the `MathFunction`s can share it.

**Review by jimblandy — CHANGES_REQUESTED** [review]

File: src/front/glsl/constants.rs
Comment (sjinno) [reply src/front/glsl/constants.rs]: Done, though I haven't handled vector arguments yet.

File: src/front/glsl/constants.rs
Comment (sjinno) [reply src/front/glsl/constants.rs]: Done.

File: src/front/glsl/constants.rs
Comment (JCapucho) [inline src/front/glsl/constants.rs:237]: nit: Can we add just a comment with something like
```rust
// TODO: when the MSRV is higher or equal to 1.50 switch to f32::clamp
```

Comment (JCapucho) [inline src/front/glsl/constants.rs:237]: We need to figure out how the `NaN`s work with this. The glsl spec says
> min returns the minimum of the two parameters. It returns y if y is less than x, otherwise it returns x. 

while the rust standard library has

> If one of the arguments is NaN, then the other argument is returned.

I couldn't find how glsl defines the float ordering with `NaN`s, but standard `IEEE 754-2008` floats don't have a stable ordering between `NaN`s and any other number.

**Review by JCapucho — CHANGES_REQUESTED** [review]
I have some concerns related to `NaN`s handling, that I would like addressed

File: src/front/glsl/constants.rs
Comment (sjinno) [reply src/front/glsl/constants.rs:237]: Yes.

Comment (sjinno) [reply src/front/glsl/constants.rs:237]: I'm so sorry for the delay. I've been meaning to get back to this, but I got a little too busy... But I am meaning to resume this unresolved issue, so I was wondering if there was any update on how the `NaN`s work with clamp. In the meantime, I'll also try digging into what I can find :)

Comment (JCapucho) [reply src/front/glsl/constants.rs:237]: Sorry at the time I didn't realize `clamp` panicked on `NaN` so we can't really use it.

Comment (JCapucho) [reply src/front/glsl/constants.rs:237]: I did found out how it worked, for `min` thanks to it's definition it's behavior is the same as rust. but for `max` this isn't the case so you'll need the following function
```rust
/// Helper function to implement the glsl `max` function for floats
///
/// While rust does provide a `f32::max` method it has a different behavior than the
/// glsl `max` for `NaN`s, in rust if any of the arguments is a `NaN` then the other
/// is returned.
///
/// This leads to different results in the following example
/// ```
/// max(NaN, 1.0)
/// ```
///
/// Rust will return 1.0 while glsl should return NaN
fn glsl_float_max(x: f32, y: f32) -> f32 {
    if x < y {
        y
    } else {
        x
    }
}
```

Comment (JCapucho) [reply src/front/glsl/constants.rs:237]: Wait sorry used some old notes, `min` is not equal between the two. you also need this function
```rust
/// Helper function to implement the glsl `min` function for floats
///
/// While rust does provide a `f32::min` method it has a different behavior than the
/// glsl `min` for `NaN`s, in rust if any of the arguments is a `NaN` then the other
/// is returned.
///
/// This leads to different results in the following example
/// ```
/// min(NaN, 1.0)
/// ```
///
/// Rust will return 1.0 while glsl should return NaN
fn glsl_float_min(x: f32, y: f32) -> f32 {
    if y < x {
        y
    } else {
        x
    }
}
```

Comment (sjinno) [reply src/front/glsl/constants.rs:237]: Thanks so much for this! I'm almost ready to push the new change, but I am wondering in which file I should be writing that function? `glsl/functions.rs`?

Comment (JCapucho) [reply src/front/glsl/constants.rs:237]: No it should be added in the same file since it concerns constant expression functions and not runtime functions

Comment (sjinno) [reply src/front/glsl/constants.rs:237]: Okay, thank you very much!

Comment (sjinno) [reply src/front/glsl/constants.rs]: I just pushed a new commit. Let me know if it works :) And thank you very much for your assist!
One thing maybe I should have gotten clarity on is that I changed `f32` to `f64` in the `glsl_float_min()` function, since `ScalarValue::Float` takes `f64`.

Comment (JCapucho) [reply src/front/glsl/constants.rs]: You forgot to also use the `max` modified variant

Comment (sjinno) [reply src/front/glsl/constants.rs]: Hopefully fixed it :)

Comment (JCapucho) [reply src/front/glsl/constants.rs]: Sorry didn't notice at first but you need to use `f64`s, when I posted the code I forgot that naga uses `f64`s

Comment (sjinno) [reply src/front/glsl/constants.rs]: Oh, no worries! That's what I was thinking too, but forgot to ask for clarity on it.

**Review by JCapucho — APPROVED** [review]
LGTM, just need to wait on @jimblandy approval

File: src/front/glsl/constants.rs
Comment (jimblandy) [inline src/front/glsl/constants.rs:564]: Capitalize "Rust", and write "GLSL" in all caps. Also, this is a run-on sentence; please change ", in rust if" to ". In Rust, if"

File: src/front/glsl/constants.rs
Comment (jimblandy) [inline src/front/glsl/constants.rs:572]: Missing period at end of sentence.

File: src/front/glsl/constants.rs
Comment (jimblandy) [inline src/front/glsl/constants.rs:581]: Same comments for `glsl_float_min` as `glsl_float_max`.

**Review by jimblandy — APPROVED** [review]
Just some minor comments - other than that, this looks good.

**Pull request merged**
